src/main/java/net/neoforged/neoforge/fluids/crafting/UnitAmountFluidIngredient.java

```diff
@@ -0,0 +1,267 @@
+            FluidIngredient.CODEC_NON_EMPTY.fieldOf("ingredient").forGetter(UnitAmountFluidIngredient::ingredient),
+            NeoForgeExtraCodecs.optionalFieldAlwaysWrite(AMOUNTS_CODEC, "amount", DEFAULT_AMOUNT).forGetter(UnitAmountFluidIngredient::amounts))
+            .apply(instance, UnitAmountFluidIngredient::new));
+
+    public static final StreamCodec<RegistryFriendlyByteBuf, UnitAmountFluidIngredient> STREAM_CODEC = StreamCodec.composite(
+            FluidIngredient.STREAM_CODEC,
+            UnitAmountFluidIngredient::ingredient,
+            ByteBufCodecs.<ByteBuf, FluidUnit, Integer, Map<FluidUnit, Integer>>map(HashMap::new, FluidUnit.STREAM_CODEC, ByteBufCodecs.INT).apply(ByteBufCodecs.list()),
+            UnitAmountFluidIngredient::amounts,
+            UnitAmountFluidIngredient::new);
+
+    /**
+     * Helper method to create a simple unit-based ingredient that has a single unit quantity.
+     */
+    public static UnitAmountFluidIngredient of(FluidIngredient fluid, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(fluid, List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Helper method to create a simple unit-based ingredient that matches a single fluid.
+     */
+    public static UnitAmountFluidIngredient of(Fluid fluid, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.of(fluid), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Helper method to create a simple unit-based ingredient that matches a single fluid.
+     */
+    public static UnitAmountFluidIngredient of(TagKey<Fluid> tag, int amount, FluidUnit unit) {
+        return new UnitAmountFluidIngredient(FluidIngredient.tag(tag), List.of(Map.of(unit, amount)));
+    }
+
+    /**
+     * Factory method for creating any ingredient.
+     */
+    @SafeVarargs
+    public static UnitAmountFluidIngredient of(FluidIngredient ingredient, Map<FluidUnit, Integer>... amounts) {
+        if (amounts.length == 0)
+            return new UnitAmountFluidIngredient(ingredient, DEFAULT_AMOUNT);
+        else
+            return new UnitAmountFluidIngredient(ingredient, Arrays.stream(amounts)
+                    .map(UnitAmountFluidIngredient::validateNotEmpty)
+                    .map(DataResult::getOrThrow).toList());
+    }
+
+    private final FluidIngredient ingredient;
+    private final List<Map<FluidUnit, Integer>> amounts;
+
+    @Nullable
+    private FluidStack[] cachedStacks;
+
+    private UnitAmountFluidIngredient(FluidIngredient ingredient, List<Map<FluidUnit, Integer>> amounts) {
+        this.ingredient = ingredient;
+        this.amounts = amounts;
+    }
+
+    public FluidIngredient ingredient() {
+        return ingredient;
+    }
+
+    public List<Map<FluidUnit, Integer>> amounts() {
+        return amounts;
+    }
+
+    public Optional<Integer> amount(FluidStack stack) {
```

Comment:
Small nitpick but why `Optional<Integer>` when `Optionalnt` exists?

Reply:
OptionalInt does not support the map/orElse etc APIs.